### PR TITLE
7615/refactor/lazybookcard cleanup requests

### LINKED ISSUE #7615: Cleanup/convert barcodescanner to use Vue
This is becoming interactive enough where having it be in Vue would be helpful.

Steps (two separate PRs):

- [ ] Update LazyBookCard so that after it fetches `/isbn/...`, it fetches `https://openlibrary.org/search.json?q=edition_key:OL38222427M&fields=key,author_name` in stead of making multiple requests for the author names.
- [ ] Create a new BarcodeScanner Vue component. Most of the helper classes should remain the same and just be moved over (eg `OCRScanner`). `LazyBookCard` should be made into a Vue component. `OLBarcodeScanner` should be made into a vue component. Life the html from `barcode-scanner.html`, and replace it with a Vue app rendering (eg `$:render_component('MergeUI', attrs=attrs)`)

### Stakeholders
@Eds-Dbug

### PR DESCRIPTION
**Pull request by leanderoudakkertue**
Closes #7615 (First checklist item)

Refactored `LazyBookCard` so that after it fetches `/isbn/...`, it fetches `/search.json?q=edition_key:{key}&fields=key,author_name` instead of making multiple requests for the author names.

### Testing

For the user, the behavior remains the same. I tested if the `/barcodescanner` page recognizes books and correctly fetches and displays the name, ISBN, authors, and cover. I also tested if the page properly redirects when the page has a `returnTo` parameter `/barcodescanner?returnTo=/isbn/$$$$`. This has been tested using different books with differing number of authors.

### Screenshot

image

### Stakeholders

@cdrini

### CODE REVIEW

File: openlibrary/plugins/openlibrary/js/page-barcodescanner/LazyBookCard.js [hunk 1]
```diff
@@ -105,16 +103,14 @@ export default class LazyBookCard {
...
-                return Promise.all(
-                    workRecord.authors
-                        .map(a => fetch(`${OL_BASE}${a.author.key}.json`).then(r => r.json()))
-                );
-            }).then(authorRecords => {
+                // Split up key, for example /book/00000 becomes 00000.
+                const edition_key = editionRecord.key.split('/').pop();
+
+                return fetch(`${OL_BASE}search.json?q=edition_key:${edition_key}&fields=key,author_name`).then(r => r.json())
+            }).then(searchRecords => {
```
Comment (cdrini) [inline openlibrary/plugins/openlibrary/js/page-barcodescanner/LazyBookCard.js:110]: Nice work @leanderoudakkertue this looks great! I just realized there's an edge case I missed when I created this issue; sometimes `/isbn` creates new books, and it takes ~a minute for them to appear in search. We don't leave enough time in that case. Can you add a fall back in here? Something like:

```
if (!searchRecords.docs.length) {
    // Do the previous more expensive method of fetching all the author records
}
```

That should let us handle both cases.

Comment (leanderoudakkertue) [reply openlibrary/plugins/openlibrary/js/page-barcodescanner/LazyBookCard.js:116]: @cdrini I implemented and tested the fallback.

**Comment by leanderoudakkertue** [comment]
I am not sure why the issue has been closed, since I didn't delete the repository (and I don't see any suspicious activity in my account). But I have contacted GitHub support to check if the PR can be reopened. Otherwise, I will make sure to create a new PR and link it back to this one.

**Comment by leanderoudakkertue** [comment]
> I am not sure why the issue has been closed, since I didn't delete the repository (and I don't see any suspicious activity in my account). But I have contacted GitHub support to check if the PR can be reopened. Otherwise, I will make sure to create a new PR and link it back to this one.

@cdrini According to GitHub support, you should be able to reopen the PR.

**Comment by cdrini** [comment]
I think you will have to make a new PR; the reopen button is disabled for me! Very strange...
Screenshot_20240503-223051

**Pull request closed without merging**